File: src/main/java/com/wowo/wowo/domain/shared/valueobjects/Money.java
package com.wowo.wowo.domain.shared.valueobjects;

import jakarta.persistence.Embeddable;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;

/**
 * Value object representing a monetary amount.
 * Encapsulates business rules for money handling.
 */
@Embeddable
@Getter
@EqualsAndHashCode
@NoArgsConstructor
public class Money {
    
    private Long amount;
    private String currency;
    
    public Money(Long amount, String currency) {
        if (amount == null) {
            throw new IllegalArgumentException("Amount cannot be null");
        }
        if (currency == null || currency.trim().isEmpty()) {
            throw new IllegalArgumentException("Currency cannot be null or empty");
        }
        this.amount = amount;
        this.currency = currency.toUpperCase();
    }
    
    public static Money zero(String currency) {
        return new Money(0L, currency);
    }
    
    public static Money vnd(Long amount) {
        return new Money(amount, "VND");
    }
    
    public Money add(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot add money with different currencies");
        }
        return new Money(this.amount + other.amount, this.currency);
    }
    
    public Money subtract(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot subtract money with different currencies");
        }
        return new Money(this.amount - other.amount, this.currency);
    }
    
    public boolean isGreaterThan(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot compare money with different currencies");
        }
        return this.amount > other.amount;
    }
    
    public boolean isGreaterThanOrEqual(Money other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException("Cannot compare money with different currencies");
        }
        return this.amount >= other.amount;
    }
    
    public boolean isNegative() {
        return amount < 0;
    }
    
    public boolean isPositive() {
        return amount > 0;
    }
    
    public boolean isZero() {
        return amount.equals(0L);
    }
    
    @Override
    public String toString() {
        return String.format("%d %s", amount, currency);
    }
}

File: src/main/kotlin/com/wowo/wowo/shared/valueobject/Money.kt
package com.wowo.wowo.shared.valueobject

import com.wowo.wowo.shared.domain.ValueObject
import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.jvm.JvmStatic

/**
 * Value Object representing monetary amount with currency
 */
data class Money(
    val amount: BigDecimal,
    val currency: Currency
) : ValueObject {

    init {
        require(amount.scale() <= 2) { "Amount can have at most 2 decimal places" }
    }

    fun add(other: Money): Money {
        require(currency == other.currency) { "Cannot add money with different currencies" }
        return Money(amount.add(other.amount).setScale(2, RoundingMode.HALF_UP), currency)
    }

    fun subtract(other: Money): Money {
        require(currency == other.currency) { "Cannot subtract money with different currencies" }
        return Money(amount.subtract(other.amount).setScale(2, RoundingMode.HALF_UP), currency)
    }

    fun multiply(multiplier: BigDecimal): Money {
        return Money(amount.multiply(multiplier).setScale(2, RoundingMode.HALF_UP), currency)
    }

    fun isPositive(): Boolean = amount > BigDecimal.ZERO
    fun isNegative(): Boolean = amount < BigDecimal.ZERO
    fun isZero(): Boolean = amount.compareTo(BigDecimal.ZERO) == 0

    companion object {
        @JvmStatic
        fun zero(currency: Currency) = Money(BigDecimal.ZERO.setScale(2), currency)
    }
}
